org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/HiddenRegionPartAssociator.java

```diff
@@ -0,0 +1,68 @@
+/*******************************************************************************
+ * Copyright (c) 2018 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.formatting2.regionaccess.internal;
+
+import java.util.List;
+import java.util.function.BiConsumer;
+
+import org.eclipse.xtext.formatting2.regionaccess.HiddenRegionPartAssociation;
+import org.eclipse.xtext.formatting2.regionaccess.IHiddenRegion;
+import org.eclipse.xtext.formatting2.regionaccess.IHiddenRegionPart;
+import org.eclipse.xtext.formatting2.regionaccess.IHiddenRegionPartAssociator;
+import org.eclipse.xtext.formatting2.regionaccess.IWhitespace;
+
+/**
+ * @author Moritz Eysholdt - Initial contribution and API
+ */
+public class HiddenRegionPartAssociator implements IHiddenRegionPartAssociator {
+
+	@Override
+	public void associate(IHiddenRegion region, BiConsumer<IHiddenRegionPart, HiddenRegionPartAssociation> handler) {
+		List<IHiddenRegionPart> parts = region.getParts();
+		if (!region.isMultiline()) {
+			for (IHiddenRegionPart p : parts) {
```

Comment:
Why don't you extract smaller methods that are easier to customize?

Reply:
because it's part of an "internal" package and therefore not intended as API. But besides this I'd rather have people copy the method impl and customize it. This makes it much easier for us to keep our API stable. A decomposition of the algorithm such as `collectLeading()` and `collectTrailing()` stops being useful once you want to have some magical "it's only leading when it's not trailing" logic.